libc/string_cpy.c

```diff
@@ -0,0 +1,1451 @@
+	TEST_ASSERT_NOT_EQUAL_CHAR(testStr[sizeof(testStr) / 2], buff[sizeof(testStr) / 2]);
+
+	memset(buff, 0, sizeof(buff));
+	memset(testStr, 0, sizeof(testStr));
+
+	/* Checking if we can copy place where NULL element is present */
+	TEST_ASSERT_EQUAL_PTR(buff, strncpy(buff, "", sizeof(buff)));
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(buff, testStr, sizeof(testStr));
+}
+
+
+TEST(string_strncpy, adjacent)
+{
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
+
+	char testStr[] = "TEST",
+		 memStr[MAX_STR_LEN] = "\0\0\0\0\0\0\0\0\0\0TEST",
+		 expVal[MAX_STR_LEN] = "\0\0\0\0\0\0TESTTEST",
+		 testStrLen = sizeof(testStr) - 1;
+
+	/* Copy data in same space but with offset to left */
+	TEST_ASSERT_EQUAL_STRING(&expVal[10 - testStrLen], strncpy(&memStr[10 - testStrLen], &memStr[10], testStrLen));
+
+	strcat(&expVal[10], testStr);
+
+	/* Copy data in same space but with offset to right */
+	TEST_ASSERT_EQUAL_STRING(&expVal[10 + testStrLen], strncpy(&memStr[10 + testStrLen], &memStr[10], testStrLen));
+
+#pragma GCC diagnostic pop
+}
+
+
+TEST(string_strncpy, one_byte)
+{
+	char buff[MAX_ASCII_SET] = { 0 },
+		 ascii[MAX_ASCII_SET] = { 0 };
+	int i;
+
+	/* Copy one by one from created ascii table */
+	for (i = 1; i < MAX_ASCII_SET; i++) {
+		ascii[i - 1] = i;
+
+		TEST_ASSERT_EQUAL_CHAR_ARRAY(buff, strncpy(buff, &ascii[i], 1), sizeof(buff));
+	}
+}
+
+
+TEST(string_strncpy, big)
+{
+	int num = 12345678;
+	char payload[BIG_NUMB] = { 0 },
+		 bigBuff[BIG_NUMB] = { 0 };
+
+	sprintf(payload, "%d", num);
+
+/* Pragma truncation disable for the length of string tests */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
+
+	/* Ability to copy long string */
+	TEST_ASSERT_EQUAL_STRING(bigBuff, strncpy(bigBuff, longStr, BIG_NUMB));
+	TEST_ASSERT_EQUAL_STRING(longStr, bigBuff);
+
+	/* Checking if a string doesn't contain any of the data after the size of input */
+	TEST_ASSERT_EQUAL_STRING(bigBuff, strncpy(bigBuff, payload, BIG_NUMB));
+	TEST_ASSERT_EQUAL_STRING(payload, bigBuff);
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(bigBuff, payload, sizeof(bigBuff));
+
+	/* Input bigger than target */
+	memset(payload, 0, sizeof(payload));
+	memset(bigBuff, 'a', sizeof(bigBuff));
+
+	TEST_ASSERT_EQUAL_PTR(payload, strncpy(payload, bigBuff, sizeof(payload)));
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(payload, &bigBuff[sizeof(bigBuff) - sizeof(payload)], sizeof(payload));
+
+#pragma GCC diagnostic pop
+}
+
+
+TEST(string_strncpy, append_null)
+{
+	char input[BIG_NUMB / 2] = { 0 }, zeroHolder[BIG_NUMB / 2] = { 0 }, buff[BIG_NUMB] = { 0 };
+	int i;
+
+	big_string();
+
+	for (i = 0; i < BIG_NUMB / 2 - 1; i++) {
+		if (longStr[i] == 0) {
+			input[i] = longStr[i + 1];
+		}
+		else {
+			input[i] = longStr[i];
+		}
+	}
+
+	/* stpncpy will search for end of array by looking for nul term */
+	input[i] = '\0';
+
+	TEST_ASSERT_EQUAL_PTR(buff, strncpy(buff, input, sizeof(buff)));
```

Comment:
You don't need buff there, you can use the second half of longStr:

```suggestion
	TEST_ASSERT_EQUAL_PTR(buff, strncpy(buff, &longStr[BIG_NUMB / 2], BIG_NUMB / 2));
```

Reply:
`input` and `sizeof(buff)` are used because input holds only chars that are not null terms and buff size is 2 times bigger than the input size